spec/services/file_hierarchy_service_spec.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe FileHierarchyService do
+  subject(:root_directory) { described_class.to_hierarchy(work_version:) }
+
+  let(:paths) { ['dir1/dir2/file2.pdf', 'dir1/file1.pdf', 'dir1/file3.pdf', 'dir1/file4.pdf', 'file5.pdf'] }
+
+  let(:attached_file1) { create(:attached_file, path: paths[0]) }
+  let(:attached_file2) { create(:attached_file, path: paths[1]) }
+  let(:attached_file3) { create(:attached_file, path: paths[2]) }
+  let(:attached_file4) { create(:attached_file, path: paths[3]) }
+  let(:attached_file5) { create(:attached_file, path: paths[4]) }
+
+  let(:work_version) do
+    build(:work_version, attached_files:
+    [attached_file1, attached_file2, attached_file3, attached_file4, attached_file5])
+  end
+
+  before do
+    allow(attached_file1).to receive(:filename).and_return(paths[0])
+    allow(attached_file2).to receive(:filename).and_return(paths[1])
+    allow(attached_file3).to receive(:filename).and_return(paths[2])
+    allow(attached_file4).to receive(:filename).and_return(paths[3])
+    allow(attached_file5).to receive(:filename).and_return(paths[4])
+  end
+
+  # rubocop:disable Layout/FirstArrayElementIndentation
+  it 'returns a hash of the file hierarchy' do
+    expect(root_directory).to match(
+      FileHierarchyService::Directory.new('', [
+        FileHierarchyService::Directory.new('dir1', [
+          FileHierarchyService::Directory.new('dir2', [], [
+            FileHierarchyService::File.new(attached_file1)
+          ], 2)
+        ], [
+          FileHierarchyService::File.new(attached_file2),
+          FileHierarchyService::File.new(attached_file3),
+          FileHierarchyService::File.new(attached_file4)
+        ], 1)
+      ], [
+        FileHierarchyService::File.new(attached_file5)
+         ], 0)
+    )
+  end
+  # rubocop:enable Layout/FirstArrayElementIndentation
+end
```

Comment:
disabled this rubocop rule here as it wanted to indent this structure to a ridiculous degree, which it less readable in my view

Reply:
WFM